### PR TITLE
feature/rejoin

### PR DESCRIPTION
**Pull request by CoryBorek**

### CODE REVIEW

**Review by thetayloredman — CHANGES_REQUESTED** [review]
Please undo your changes to line endings -- they should be LF. This is probably an issue with your editor or Git configuration. I can fix this for you if you'd like

**Comment by thetayloredman** [comment]
PR #6 fixes line endings.

**Comment by ilikestohack** [comment]
Also I do not believe this will be merged into master... Master was originally the branch for the actual public plugin version so not sure how we want to do this

**Comment by thetayloredman** [comment]
> Also I do not believe this will be merged into master... Master was originally the branch for the actual public plugin version so not sure how we want to do this

@ilikestohack Main would be the correct branch to merge this into. It's a fork, so the histories are separate like that

**Comment by ilikestohack** [comment]
> > Also I do not believe this will be merged into master... Master was originally the branch for the actual public plugin version so not sure how we want to do this
> 
> @ilikestohack Main would be the correct branch to merge this into. It's a fork, so the histories are separate like that

Well no I mean the forks original purpose was to contribute upstream so I feel like we should not break that functionality

**Comment by thetayloredman** [comment]
> Well no I mean the forks original purpose was to contribute upstream so I feel like we should not break that functionality

Merge it to a new branch and do a cherry-pick onto main?

File: Core/src/main/java/com/theprogrammingturkey/comz/game/Game.java [hunk 1]
```diff
@@ -58,11 +53,7 @@
 	/**
 	 * List of every player contained in game.
 	 */
-	public List<Player> players = new ArrayList<>();
-	/**
-	 * List of all the spectating players contained in game.
-	 */
-	public List<Player> spectators = new ArrayList<>();
+	public Map<Player,GamePlayer> gamePlayers = new LinkedHashMap<>();
```
Comment (thetayloredman) [inline Core/src/main/java/com/theprogrammingturkey/comz/game/Game.java:56]:
```suggestion
	public Map<Player, GamePlayer> gamePlayers = new LinkedHashMap<>();
```

File: Core/src/main/java/com/theprogrammingturkey/comz/game/Game.java [hunk 2]
```diff
@@ -268,7 +254,20 @@ public Game(String name)
...
+		return null;
+	}
+
+	public List<Player> getPlayers() {
+		List<Player> players = new ArrayList<>();
+		for (GamePlayer value : gamePlayers.values()) {
+			if (value.getState().equals(GamePlayer.PlayerState.IN_GAME)) players.add(value.getPlayer());
+		}
+		return players;
 	}
```
Comment (thetayloredman) [inline Core/src/main/java/com/theprogrammingturkey/comz/game/Game.java:271]:
```suggestion
	public List<Player> getPlayers() {
		List<Player> players = new ArrayList<>(gamePlayers.values());
		players.removeIf(v -> !v.getState().equals(GamePlayer.PlayerState.IN_GAME));
		return players;
	}
```

I don't know Java, but this looks about right for how it should be in most langs (the removeif call should work)
**DO NOT APPLY BLINDLY.** Test first.

File: Core/src/main/java/com/theprogrammingturkey/comz/game/Game.java [hunk 3]
```diff
@@ -268,7 +254,20 @@ public Game(String name)
...
 	public PlayerWeaponManager getPlayersWeapons(Player player)
 	{
-		return playersGuns.computeIfAbsent(player, PlayerWeaponManager::new);
+		for (Player pl : getPlayers()) {
+			if (pl.equals(player)) {
+				return gamePlayers.get(player).getWeaponManager();
+			}
+		}
+		return null;
+	}
```
Comment (thetayloredman) [inline Core/src/main/java/com/theprogrammingturkey/comz/game/Game.java:263]:
```suggestion
		Player pl = getPlayers().stream().filter(pl -> pl.equals(player)).findAny().orElse(null);
		if (pl == null) return pl;
		return gamePlayers.get(pl).getWeaponManager();
	}
```

**DO NOT APPLY BLINDLY.** This needs testing.

File: Core/src/main/java/com/theprogrammingturkey/comz/game/Game.java [hunk 4]
```diff
@@ -1326,24 +1390,32 @@ public void updateBarrierDamage(int damage, Collection<Block> blocks)
...
-		ArrayList<Player> combined = new ArrayList<>(players);
-		combined.addAll(spectators);
-		return combined;
+		List<Player> out = new ArrayList<>();
+		for (GamePlayer value : gamePlayers.values()) {
+			if (value.getState().equals(GamePlayer.PlayerState.IN_GAME) || value.getState().equals(GamePlayer.PlayerState.SPECTATING)) {
+				out.add(value.getPlayer());
+			}
+		}
+		return out;
```
Comment (thetayloredman) [inline Core/src/main/java/com/theprogrammingturkey/comz/game/Game.java:1413]: Should use a filter/removeIf here instead.

**Review by thetayloredman — CHANGES_REQUESTED** [review]

File: Core/src/main/java/com/theprogrammingturkey/comz/commands/CommandManager.java [hunk 5]
```diff
@@ -1,278 +1,279 @@
...
+		commandList.put("leaders", commandList.get("leaderboard"));
+
+		// Leave the given arena:
+		// Perm
+		// - zombies.leave
+		// - zombies.user
+		commandList.put("leave", new LeaveCommand());
+		commandList.put("l", commandList.get("leave"));
+
+		commandList.put("rejoin", new RejoinCommand());
```
[the remark this answers is not in the document]
Comment (thetayloredman) [reply Core/src/main/java/com/theprogrammingturkey/comz/commands/CommandManager.java]: Not needed anymore

File: Core/src/main/java/com/theprogrammingturkey/comz/game/Game.java [hunk 6]
```diff
@@ -1,1381 +1,1453 @@
...
+	 * Gets the weapons the player currently has
+	 *
+	 * @param player to get the weapons of
+	 * @return PlayerWeaponManager of the players weapons
+	 */
+	public PlayerWeaponManager getPlayersWeapons(Player player)
+	{
+		for (Player pl : getPlayers()) {
+			if (pl.equals(player)) {
+				return gamePlayers.get(player).getWeaponManager();
```
[the remark this answers is not in the document]
Comment (thetayloredman) [reply Core/src/main/java/com/theprogrammingturkey/comz/game/Game.java]: Superseded

File: Core/src/main/java/com/theprogrammingturkey/comz/game/Game.java (hunk 3 above)
Comment (CoryBorek) [reply Core/src/main/java/com/theprogrammingturkey/comz/game/Game.java:263]: That won't function, it wants a PlayerWeaponManager, not a player, so if it's null, it won't function

Comment (CoryBorek) [reply Core/src/main/java/com/theprogrammingturkey/comz/game/Game.java:263]: Oh wait, that's if player is null, could be a little confusing

File: Core/src/main/java/com/theprogrammingturkey/comz/commands/RejoinCommand.java [hunk 7]
```diff
@@ -14,19 +16,19 @@ public class RejoinCommand implements SubCommand
...
 			CommandUtil.sendMessageToPlayer(player, ChatColor.RED + "There is no game to rejoin!");
 			return true;
 		}
-
-		if (game.gamePlayers.get(player).getState().equals(GamePlayer.PlayerState.LEFT_GAME)) {
-			game.addPlayer(player);
-		}
-		else {
-			CommandUtil.sendMessageToPlayer(player, ChatColor.RED + "You are already in the game!");
+		if (COMZPermission.REJOIN_AREMA.hasPerm(player, game.getName())) {
```
Comment (thetayloredman) [inline Core/src/main/java/com/theprogrammingturkey/comz/commands/RejoinCommand.java:24]: AREMA

File: Core/src/main/java/com/theprogrammingturkey/comz/util/COMZPermission.java [hunk 8]
```diff
@@ -7,6 +7,7 @@
 public enum COMZPermission
 {
 	JOIN_ARENA(true, "zombies.join", "zombies.join.%%"),
+	REJOIN_AREMA(true, "zombies.rejoin", "zombies.rejoin.%%"),
```
Comment (thetayloredman) [inline Core/src/main/java/com/theprogrammingturkey/comz/util/COMZPermission.java:10]: AREMA II

File: Core/src/main/java/com/theprogrammingturkey/comz/commands/InfoCommand.java [hunk 9]
```diff
@@ -45,9 +48,9 @@ else if(GameManager.INSTANCE.isValidArena(args[1]))
 					CommandUtil.sendMessageToPlayer(player, ChatColor.GREEN + "Lobby Spawn: x:" + ChatColor.BLUE + game.getLobbyLocation().getBlockX() + ChatColor.GREEN + ", y:" + ChatColor.BLUE + game.getLobbyLocation().getBlockY() + ChatColor.GREEN + ", z:" + ChatColor.BLUE + game.getLobbyLocation().getBlockZ());
 					CommandUtil.sendMessageToPlayer(player, ChatColor.GREEN + "Spectator Spawn: x:" + ChatColor.BLUE + game.getSpectateLocation().getBlockX() + ChatColor.GREEN + ", y:" + ChatColor.BLUE + game.getSpectateLocation().getBlockY() + ChatColor.GREEN + ", z:" + ChatColor.BLUE + game.getSpectateLocation().getBlockZ());
 					CommandUtil.sendMessageToPlayer(player, ChatColor.GREEN + "Players: ");
-
-					for(int i = 0; i < game.players.size(); i++)
-						CommandUtil.sendMessageToPlayer(player, ChatColor.BLUE + "  " + game.players.get(i).getName());
+					List<Player> players = game.getPlayers();
+					for(int i = 0; i < players.size(); i++)
+						CommandUtil.sendMessageToPlayer(player, ChatColor.BLUE + "  " + game.getPlayers().get(i).getName());
```
Comment (thetayloredman) [inline Core/src/main/java/com/theprogrammingturkey/comz/commands/InfoCommand.java:53]:
```suggestion
						CommandUtil.sendMessageToPlayer(player, ChatColor.BLUE + "  " + players.get(i).getName());
```

File: Core/src/main/java/com/theprogrammingturkey/comz/commands/JoinCommand.java [hunk 10]
```diff
@@ -3,18 +3,26 @@
...
+import java.util.Map;
+
 public class JoinCommand implements SubCommand
 {
 	@Override
 	public boolean onCommand(Player player, String[] args)
 	{
 		if(GameManager.INSTANCE.isPlayerInGame(player))
 		{
+			Map<Player,GamePlayer> gamePlayers = GameManager.INSTANCE.getGame(player).gamePlayers;
```
Comment (thetayloredman) [inline Core/src/main/java/com/theprogrammingturkey/comz/commands/JoinCommand.java:21]:
```suggestion
			Map<Player, GamePlayer> gamePlayers = GameManager.INSTANCE.getGame(player).gamePlayers;
```

**Review by thetayloredman — CHANGES_REQUESTED** [review]
This is just high-level stylistic stuff, I don't know Java and I don't know the project well enough to do anything better. I'm assuming this is being tested

**Comment by CoryBorek** [comment]
> This is just high-level stylistic stuff, I don't know Java and I don't know the project well enough to do anything better. I'm assuming this is being tested

Yeah, it's all tested and functional.

**Review by thetayloredman — APPROVED** [review]
I can't see anything wrong here. I haven't done a control flow analysis or anything like that but overall LGTM! CC for review & merge: @ilikestohack

**Comment by thetayloredman** [comment]
r+
r? ilikestohack

**Pull request merged**